sql/tests: TestDeadlockDetection failed

sql/tests.TestDeadlockDetection [failed](https://tanzanite.cluster.engflow.com/invocations/default/cb2700f9-6b93-44d9-a17c-a8295df892e0?testReportRun=2&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cvc3FsL3Rlc3RzOnRlc3RzX3Rlc3Q=) on master @ [7211211793b70b1759874f8b53613a9c9682270d](https://github.com/cockroachdb/cockroach/commits/7211211793b70b1759874f8b53613a9c9682270d):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xa0c1981]

```
Stack: 

```
goroutine 13 [running]:
testing.tRunner.func1.2({0xab67e20, 0x11795c50})
	GOROOT/src/testing/testing.go:1632 +0x3fc
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x6b6
panic({0xab67e20?, 0x11795c50?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:136 +0x51a
panic({0xab67e20?, 0x11795c50?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc005296ac8?, {0xcea1ee0, 0x12f3e280})
	pkg/util/stop/stopper.go:226 +0x74
panic({0xab67e20?, 0x11795c50?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc00406f040, {0xcea1ee0, 0x12f3e280})
	pkg/util/stop/stopper.go:566 +0x2c5
panic({0xab67e20?, 0x11795c50?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/sql/tests.TestDeadlockDetection(0xc003f9b1e0)
	pkg/sql/tests/deadlock_test.go:132 +0xdc1
testing.tRunner(0xc003f9b1e0, 0xbc01e28)
	GOROOT/src/testing/testing.go:1690 +0x227
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1743 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestDeadlockDetection
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestDeadlockDetection3056138556
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
*
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:791
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop
*   | 	pkg/util/stop/stopper.go:566
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:791
*   | runtime.panicmem
*   | 	GOROOT/src/runtime/panic.go:262
*   | runtime.sigpanic
*   | 	GOROOT/src/runtime/signal_unix.go:917
*   | github.com/cockroachdb/cockroach/pkg/sql/tests.TestDeadlockDetection
*   | 	pkg/sql/tests/deadlock_test.go:132
*   | testing.tRunner
*   | 	GOROOT/src/testing/testing.go:1690
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1700
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
*
    stopper.go:226: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestDeadlockDetection3056138556
    stopper.go:226: panic: runtime error: invalid memory address or nil pointer dereference
--- FAIL: TestDeadlockDetection (12.58s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDeadlockDetection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50932

--- a/pkg/sql/tests/deadlock_test.go
+++ b/pkg/sql/tests/deadlock_test.go
@@ -50,6 +50,12 @@ func TestDeadlockDetection(t *testing.T) {
 	defer srv.Stopper().Stop(ctx)
 	sqlr := sqlutils.MakeSQLRunner(db)
 
+	hostRunner := sqlutils.MakeSQLRunner(srv.SystemLayer().SQLConn(t))
+	// If we happen to enable buffered writes metamorphically, we must have the
+	// split lock reliability enabled (which can be tweaked metamorphically too,
+	// #147217).
+	hostRunner.Exec(t, "SET CLUSTER SETTING kv.lock_table.unreplicated_lock_reliability.split.enabled = true")
+
 	sqlr.Exec(t, "CREATE TABLE test1 (id int, age int)")
 	sqlr.Exec(t, "CREATE TABLE test2 (id int, age int)")
 	sqlr.Exec(t, "INSERT INTO test1 (id, age) VALUES (1, 0)")
@@ -116,7 +122,7 @@ func TestDeadlockDetection(t *testing.T) {
 			wg.Done()
 		}()
 
-		// After executing the next operation, a deadlock is created.  Measure the
+		// After executing the next operation, a deadlock is created. Measure the
 		// time it takes for the deadlock condition to be broken.
 		_, tx1Err := tx1.Exec("UPDATE test2 SET age = 1 WHERE id = 1")
 		wg.Wait()
@@ -127,9 +133,11 @@ func TestDeadlockDetection(t *testing.T) {
 		if tx1Err != nil {
 			require.NoError(t, tx2Err)
 			require.Contains(t, tx1Err.Error(), "TransactionRetryWithProtoRefreshError")
-		} else {
+		} else if tx2Err != nil {
 			require.NoError(t, tx1Err)
 			require.Contains(t, tx2Err.Error(), "TransactionRetryWithProtoRefreshError")
+		} else {
+			t.Fatal("neither txn encountered an error")
 		}
 
 		require.NoError(t, tx1.Rollback())

